pkg/client/client_venconn_test.go

```diff
@@ -0,0 +1,381 @@
+		// to not support it because it isn't recommended.
+		given: undent(`
+			apiVersion: jetstack.io/v1alpha1
+			kind: VenafiConnection
+			metadata:
+			  name: venafi-components
+			  namespace: venafi
+			spec:
+			  vcp:
+			    url: FAKE_VENAFI_CLOUD_URL
+			    apiKey:
+			      - secret:
+			          name: apikey
+			          fields: [apikey]`),
+		expectReadyCondMsg: "b099d634ccec56556da28028743475dab67f79d079b668bedc3ef544f7eed2f3",
+		expectErr:          "VenafiConnection venafi/venafi-components: the agent cannot be used with an API key",
+	}))
+	t.Run("error when the tpp field is used", run(testcase{
+		// IMPORTANT: The user may think they can use 'tpp', spend time
+		// debugging and making the venafi connection work, and then find out
+		// that it doesn't work. The reason is because as of now, we don't first
+		// check if the user has used the 'tpp' field before running Get.
+		given: undent(`
+			apiVersion: jetstack.io/v1alpha1
+			kind: VenafiConnection
+			metadata:
+			  name: venafi-components
+			  namespace: venafi
+			spec:
+			  tpp:
+			    url: FAKE_TPP_URL
+			    accessToken:
+			      - secret:
+			          name: accesstoken
+			          fields: [accesstoken]`),
+		expectErr:          ``,
+		expectReadyCondMsg: `ea744d098c2c1c6044e4c4e9d3bf7c2a68ef30553db00f1714886cedf73230f1`,
+	}))
+}
+
+type testcase struct {
+	given              string
+	expectErr          string
+	expectReadyCondMsg string
+}
+
+func run(test testcase) func(t *testing.T) {
+	return func(t *testing.T) {
+		fakeVenafiCloud, certCloud := fakeVenafiCloud(t)
+		fakeTPP, certTPP := fakeTPP(t)
+		_, restconf, kclient := startEnvtest(t)
+
+		certPool := x509.NewCertPool()
+		certPool.AddCert(certCloud)
+		certPool.AddCert(certTPP)
+
+		cl, err := client.NewVenConnClient(
+			restconf,
+			&api.AgentMetadata{ClusterID: "no"},
+			"venafi",            // Namespace in which the Agent is running.
+			"venafi-components", // Name of the VenafiConnection.
+			"venafi",            // Namespace of the VenafiConnection.
+			certPool,
+		)
+		require.NoError(t, err)
+
+		// This `cancel` is important because the below func `Start(ctx)` needs
+		// to be stopped before the apiserver is stopped. Otherwise, the test
+		// fail with the message "timeout waiting for process kube-apiserver to
+		// stop". See:
+		// https://github.com/jetstack/venafi-connection-lib/pull/158#issuecomment-1949002322
+		// https://github.com/kubernetes-sigs/controller-runtime/issues/1571#issuecomment-945535598
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			err = cl.Start(ctx)
+			require.NoError(t, err)
+		}()
+		t.Cleanup(cancel)
+
+		// Apply the same RBAC as what you would get from the Venafi
+		// Connection Helm chart, for example after running this:
+		//  helm template venafi-connection oci://registry.venafi.cloud/charts/venafi-connection --version v0.1.0 -n venafi --show-only templates/venafi-connection-rbac.yaml
```

Comment:
Consider using the `parse` and `undent` functions and create all this content as a yaml string, which would be easier to read and easier to judge whether this *is* the same as the yaml in the Helm chart.
Optional.

Reply:
Done!